### PR TITLE
Add exchange currencies functionality

### PR DESCRIPTION
**Pull request by Cebix90**
Added functionality to handle user queries for currency value conversion using the specific exchange rate for the given day. The user provides the source currency, amount, optional date, and target currency. Proper unit tests are also implemented.

### CODE REVIEW

File: src/main/java/org/currencygoldexchangeapp/Main.java [hunk 1]
```diff
@@ -2,31 +2,43 @@
...
+            System.out.print("Enter source currency code: ");
+            String sourceCurrencyCode = scanner.next();
 
-            System.out.print("Enter date (optional, press ENTER for today's date): ");
+            System.out.print("Enter target currency code (press ENTER if not applicable): ");
+            scanner.nextLine();
+            String targetCurrencyCode = scanner.nextLine();
+
+            System.out.print("Enter amount: ");
+            double amount = scanner.nextDouble();
```
Comment (blazejprzyluski) [inline src/main/java/org/currencygoldexchangeapp/Main.java]: What happens when the user enters something else for example "test"?

File: src/main/java/org/currencygoldexchangeapp/Main.java [hunk 2]
```diff
@@ -2,31 +2,43 @@
...
 
             ExchangeRateAPIHandler exchangeRateAPIHandler = new ExchangeRateAPIHandler(HttpClient.newHttpClient());
-            CurrencyExchange exchangeRateResult = exchangeRateAPIHandler.getExchangeRateSingleCurrency(currencyCode, date.toString());
+            CurrencyExchangeCalculateService currencyExchangeCalculateService = new CurrencyExchangeCalculateService(exchangeRateAPIHandler);
 
-            System.out.println("Exchange rate for " + currencyCode + " on " + date + ": " + exchangeRateResult);
+            try {
+                CurrencyExchange result = currencyExchangeCalculateService.calculateExchangeAmount(sourceCurrencyCode, amount, targetCurrencyCode, date.toString());
+                System.out.println("Result of currency exchange: " + result);
+            } catch (Exception e) {
```
Comment (blazejprzyluski) [inline src/main/java/org/currencygoldexchangeapp/Main.java]: Why general Exception? not something more granular

File: src/main/java/org/currencygoldexchangeapp/handlers/ExchangeRateAPIHandler.java [hunk 3]
```diff
@@ -50,9 +51,11 @@ private HttpResponse<String> getHttpResponse(HttpRequest request) {
 
     private CurrencyExchange handleHttpResponse(HttpResponse<String> response) {
         if (response.statusCode() == HttpURLConnection.HTTP_OK) {
-            System.out.println("Status code: " + response.statusCode());
-            System.out.println("Response Body:");
+//            System.out.println("Status code: " + response.statusCode());
+//            System.out.println("Response Body:");
             return jsonMapper.deserializeJsonToCurrencyExchange(response.body());
+        } else if (response.statusCode() == HttpURLConnection.HTTP_NOT_FOUND) {
+            throw new CurrencyNotFoundException("The requested currency or date was not found.");
```
Comment (blazejprzyluski) [inline src/main/java/org/currencygoldexchangeapp/handlers/ExchangeRateAPIHandler.java]: CurrencyNotFound exception should only apply if the currency was not found, adding date here is confusing, don't you think?

File: src/main/java/org/currencygoldexchangeapp/services/CurrencyExchangeCalculateService.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+import java.math.RoundingMode;
+
+public class CurrencyExchangeCalculateService {
+    private final ExchangeRateAPIHandler exchangeRateAPIHandler;
+
+    public CurrencyExchangeCalculateService(ExchangeRateAPIHandler exchangeRateAPIHandler) {
+        this.exchangeRateAPIHandler = exchangeRateAPIHandler;
+    }
+
+    public CurrencyExchange calculateExchangeAmount (String sourceCurrency, double amount, String targetCurrency, String date) {
```
Comment (blazejprzyluski) [inline src/main/java/org/currencygoldexchangeapp/services/CurrencyExchangeCalculateService.java:16]: About the naming here:  what does exchange mean in this context?

Comment (Cebix90) [reply src/main/java/org/currencygoldexchangeapp/services/CurrencyExchangeCalculateService.java:16]: The name of the calculateExchangeAmount method suggests that this method deals with calculations related to the exchange of one currency for another.

In my opinion, the name of this method is quite clear, but I could make it even more precise e.g. calculateAmountInTargetCurrency

File: src/main/java/org/currencygoldexchangeapp/handlers/ExchangeRateAPIHandler.java (hunk 3 above)
Comment (Cebix90) [reply src/main/java/org/currencygoldexchangeapp/handlers/ExchangeRateAPIHandler.java]: It's true. Exception has been changed to DataNotFoundException with proper message. Now CurrencyNotFoundException is used in another place to check properly if currency is available or not.

File: src/main/java/org/currencygoldexchangeapp/Main.java (hunk 1 above)
Comment (Cebix90) [reply src/main/java/org/currencygoldexchangeapp/Main.java]: Yes i knew about this case, but I didnt put much effort in this class. After newest commit it should work properly.

File: src/main/java/org/currencygoldexchangeapp/Main.java (hunk 2 above)
Comment (Cebix90) [reply src/main/java/org/currencygoldexchangeapp/Main.java]: As above.

File: src/main/java/org/currencygoldexchangeapp/Main.java [hunk 5]
```diff
@@ -47,32 +41,73 @@ public static void main(String[] args) {
...
+
+        int year = Integer.parseInt(date.substring(0, 4));
+
+        List<String> availableCurrencies;
+
+        if (year >= 2011) {
+            availableCurrencies = codeCurrencyAvailableFor2011AndLater;
+        } else if (year >= 2003) {
+            availableCurrencies = codeCurrencyAvailableFor2010AndBefore;
+        } else {
```
Comment (blazejprzyluski) [inline src/main/java/org/currencygoldexchangeapp/Main.java:87]: Should you even accept dates outside of this range above? Please include this into the menu validation in future commit

**Review by blazejprzyluski — APPROVED** [review]
Left small comment, can be addressed later

File: src/main/java/org/currencygoldexchangeapp/Main.java (hunk 5 above)
Comment (Cebix90) [reply src/main/java/org/currencygoldexchangeapp/Main.java:87]: Yes I agree, It shouldn't be accepted. Added logic to handle these cases

**Pull request merged**